052-record-reconciled-versions.md

```diff
@@ -0,0 +1,160 @@
+
+The following is the proposed new fields:
+
+### status.versions.strimziReconciled
+The `status.versions.strimziReconciled` field is patched at the end of a **successful** reconcile, and signals to a user that the operator at version `X` was able to reconcile to successful completion.
+This would be the field a user or System Test would query against to check whether upgrade was complete.
+```
+status:
+  versions:
+    strimziReconciled: '0.37.0'
+```
+If a reconcile does not reach completion, this field is not added or updated, however in the case where it already exists it is left at the previous value.
+i.e.
+`status.versions.strimziReconciled=0.35.0` would remain if a reconcilliation on `0.36.0` failed as the information that a prior reconcilliation passed could be useful to an end user.
+
+### status.versions.strimziManagedBy
+The `status.versions.strimziManagedBy` field is patched to the CR at the start of reconcile in the `initialStatus` and signals to a user that the operator is attempting to reconcile the CR at version `Y`, regardless of success or failure.
+```
+status:
+  versions:
+    strimziManagedBy: '0.37.0'
+```
+
+This field is required for the following reasons:
+- It signals that a reconcile on version Y has started, and that the operator reconciler & watcher are functional without checking the logs of the operator.
+- The operator updates the `status.versions.strimziManagedBy` field on reconciliation startup without overriding the `status.versions.strimziReconciled` field. If the operator errors during reconcile it will be clear that the currently reconciled version does not match the version that the operator is applying. A two stage commit mechanism here gives much more information regarding whether it started the reconciliation versus whether it finished successfully.
+
+An important note here, in order to avoid unnecessary extra reconcilliations, and to avoid constant re-adding and removal of the field, this proposal suggests that `status.versions.strimziManagedBy` remains, even after a reconcile, and that the annotation is only ever updated once a new operator version attempts the reconcile.
+
+
+## `Kafka` Custom Resource changes
+- The update to `status.versions.strimziManagedBy` would happen in the `initialStatus` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java#L292)
+- The update to `status.versions.strimziReconciled` would happen in the `createOrUpdate` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java#L137) alongside the `Ready` update where the operator version would be updated into the field.
```

Comment:
You need to manage both versions in both methods. Otherwise they will be appearing and disappearing which would make them useless.